core/src/data_proof_v2.rs

```diff
@@ -1,97 +1,91 @@
-#[derive(
-	Debug, Default, Clone, Eq, Encode, Decode, PartialEq, TypeInfo, Serialize, Deserialize,
-)]
-#[serde(rename_all = "camelCase")]
-pub struct Message {
-	pub message_type: MessageType,
-	pub from: H256,
-	pub to: H256,
-	pub origin_domain: u32,
-	pub destination_domain: u32,
-	pub data: BoundedData,
-	pub id: u64, // a global nonce that is incremented with each leaf
-}
+/// Deconstructs the Unique Tx identifier into its block number and index.
+pub fn tx_uid_deconstruct(uid: u64) -> (u32, u32) {
+	const SLICE_ERR: &str = "Valid slice .qed";
 
-impl Message {
-	pub fn abi_encode(self) -> Vec<u8> {
-		encode(&[Token::Tuple(vec![
-			Token::FixedBytes(self.message_type.into()),
-			Token::FixedBytes(self.from.to_fixed_bytes().to_vec()),
-			Token::FixedBytes(self.to.to_fixed_bytes().to_vec()),
-			Token::Uint(ethabi::Uint::from(self.origin_domain)),
-			Token::Uint(ethabi::Uint::from(self.destination_domain)),
-			Token::Bytes(self.data.into()),
-			Token::Uint(ethabi::Uint::from(self.id)),
-		])])
-	}
+	let id: [u8; 8] = uid.to_be_bytes();
+	let block = u32::from_be_bytes(id[..4].try_into().expect(SLICE_ERR));
+	let tx_index = u32::from_be_bytes(id[4..].try_into().expect(SLICE_ERR));
+
+	(block, tx_index)
 }
 
-#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, Default, Serialize, Deserialize)]
-#[serde(rename_all = "camelCase")]
+#[derive(Clone, Debug, Encode, Decode, TypeInfo, Constructor)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
 pub struct ProofResponse {
 	pub data_proof: DataProofV2,
-	#[serde(skip_serializing_if = "Option::is_none")]
+	#[cfg_attr(feature = "serde", serde(skip_serializing_if = "Option::is_none"))]
 	pub message: Option<Message>,
 }
 
-#[derive(PartialEq, Debug)]
+#[derive(Debug, PartialEq, Eq)]
 pub enum SubTrie {
-	Left,
-	Right,
+	DataSubmit,
+	Bridge,
+}
+
+#[derive(Debug, Clone, Copy, Encode, Decode, PartialEq, Eq, Default, TypeInfo)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct TxDataRoots {
+	/// Global Merkle root
+	pub root: H256,
+	/// Merkle root hash of submitted data
+	pub submitted: H256,
+	/// Merkle root of bridged data
+	pub bridged: H256,
+}
+
+#[cfg(feature = "runtime")]
+impl TxDataRoots {
+	pub fn new(submitted: H256, bridged: H256) -> Self {
+		// keccak_256(submitted, bridged)
+		let sub_roots = [submitted.to_fixed_bytes(), bridged.to_fixed_bytes()].concat();
+		let root = keccak_256(sub_roots.as_slice()).into();
+
+		Self {
+			root,
+			submitted,
+			bridged,
+		}
+	}
 }
 
 /// Wrapper of `binary-merkle-tree::MerkleProof` with codec support.
-#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, Default)]
+#[derive(Clone, Debug, PartialEq, Eq, Encode, Decode, TypeInfo, Default)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
 pub struct DataProofV2 {
```

Comment:
It might be good to rename `DataProofV2` to `DataProof` and remove the old struct which is not in use since there is only one kate rpc call for data proof.